policy/pipeline/pipeline_version.rego

```diff
@@ -0,0 +1,68 @@
+package policy.pipeline.pipeline_version
+
+import data.lib
+
+VERSION_REGEX := `^[0-9]+(\.[0-0]+){0,2}`
+
+# METADATA
+# title: Pipeline defines a version
+# description: |-
+#   The Pipeline has an explicitly defined version value. The version is
+#   expected to be defined by the "app.kubernetes.io/version" resource label.
+#
+# custom:
+#   short_name: pipeline_version_exists
+#   failure_msg: Pipeline %q does not define a version
+#   effective_on: 2023-01-01T00:00:00Z
+#
+deny[result] {
+	pipeline_name := input.metadata.name
+
+	not input.metadata.labels["app.kubernetes.io/version"]
+
+	result := lib.result_helper(rego.metadata.chain(), [pipeline_name])
+}
+
+# METADATA
+# title: Pipeline version must be in a specific format
+# description: |-
+#   The Pipeline version is expected to be in one of the following formats:
+#   1
+#   1.2
+#   1.2.3
+#
+#   Each part of the version must have 1 or more digits. All other characters are
+#   not allowed.
+#
+# custom:
+#   short_name: pipeline_version_format
+#   failure_msg: Pipeline %q defines an invalid version format, %s. Use "." separated digits
+#   effective_on: 2023-01-01T00:00:00Z
+#
+deny[result] {
+	pipeline_name := input.metadata.name
+	version := input.metadata.labels["app.kubernetes.io/version"]
+	not regex.match(VERSION_REGEX, version)
+	result := lib.result_helper(rego.metadata.chain(), [pipeline_name, version])
+}
+
+# METADATA
+# title: Pipeline version is outdated
+# description: |-
+#   The Pipeline version is expected to meet a minimum required version.
+#
+# custom:
+#   short_name: pipeline_version
+#   failure_msg: Version of Pipeline %q is outdated, %s. Update to %s or newer
+#   rule_data:
+#     # TODO: Read this from another file?
```

Comment:
This needs further work.

A more sophisticated data structure is needed so we can:
1. Specify a different minimum version for specific Pipelines and Tasks, maybe by name.
2. Specify a list of minimum versions alongside with their `effective_on` date.

Maybe something like this:
```yaml
tasks:
  buildah:
    min_versions:
    - version: "0.4"
      effective_on: 2023-01-01T00:00:00Z
    - version: "0.3"
      effective_on: 2022-06-30T00:00:00Z
    - version: "0.2"
      effective_on: 2022-01-09T00:00:00Z
pipelines:
  docker-build:
    min_versions:
    - version: "2.0"
      effective_on: 2023-02-01T00:00:00Z
    - version: "1.9.2"
      effective_on: 2022-07-03T00:00:00Z
```

It's probably a good idea to have this data in an external file so it can be more easily updated by a future automated process.

Reply:
Added support for `min_versions` (2 from previous comment)